File: exercise/java/day23/src/main/java/core/control/ReindeerPowerUnit.java
package core.control;

import external.deer.Reindeer;

public class ReindeerPowerUnit {
    public Reindeer reindeer;
    public MagicPowerAmplifier amplifier = new MagicPowerAmplifier(AmplifierType.BASIC);

    public ReindeerPowerUnit(Reindeer reindeer) {
        this.reindeer = reindeer;
    }

    public float harnessMagicPower() {
        if (!reindeer.needsRest()) {
            reindeer.timesHarnessing++;
            return amplifier.amplify(reindeer.getMagicPower());
        }

        return 0;
    }

    public float checkMagicPower() {
        return reindeer.getMagicPower();
    }
}

File: exercise/scala/day23/src/main/scala/core/control/ReindeerPowerUnit.scala
package core.control

import external.deer.Reindeer

class ReindeerPowerUnit(val reindeer: Reindeer) {
  private val amplifier = new MagicPowerAmplifier(AmplifierType.BASIC)

  def harnessMagicPower(): Float = {
    if (!reindeer.needsRest()) {
      reindeer.timesHarnessing += 1
      amplifier.amplify(reindeer.getMagicPower)
    } else 0f
  }

  def checkMagicPower(): Float = reindeer.getMagicPower
}